// rookout/protobuf/variant2_pb.js
"use strict";

// source: variant2.proto

/**
 * @fileoverview
 * @enhanceable
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
var jspb = require('google-protobuf');

var goog = jspb;
var global = Function('return this')();

var google_protobuf_timestamp_pb = require('google-protobuf/google/protobuf/timestamp_pb.js');

goog.object.extend(proto, google_protobuf_timestamp_pb);

var variant_pb = require('./variant_pb.js');

goog.object.extend(proto, variant_pb);
goog.exportSymbol('proto.com.rookout.Error2', null, global);
goog.exportSymbol('proto.com.rookout.Variant2', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */

proto.com.rookout.Error2 = function (opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};

goog.inherits(proto.com.rookout.Error2, jspb.Message);

if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.com.rookout.Error2.displayName = 'proto.com.rookout.Error2';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */


proto.com.rookout.Variant2 = function (opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.com.rookout.Variant2.repeatedFields_, null);
};

goog.inherits(proto.com.rookout.Variant2, jspb.Message);

if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.com.rookout.Variant2.displayName = 'proto.com.rookout.Variant2';
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.com.rookout.Error2.prototype.toObject = function (opt_includeInstance) {
    return proto.com.rookout.Error2.toObject(opt_includeInstance, this);
  };
  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.com.rookout.Error2} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */


  proto.com.rookout.Error2.toObject = function (includeInstance, msg) {
    var f,
        obj = {
      message: jspb.Message.getFieldWithDefault(msg, 1, ""),
      type: jspb.Message.getFieldWithDefault(msg, 2, ""),
      parameters: (f = msg.getParameters()) && proto.com.rookout.Variant2.toObject(includeInstance, f),
      exc: (f = msg.getExc()) && proto.com.rookout.Variant2.toObject(includeInstance, f),
      traceback: (f = msg.getTraceback()) && proto.com.rookout.Variant2.toObject(includeInstance, f)
    };

    if (includeInstance) {
      obj.$jspbMessageInstance = msg;
    }

    return obj;
  };
}
/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.com.rookout.Error2}
 */


proto.com.rookout.Error2.deserializeBinary = function (bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.com.rookout.Error2();
  return proto.com.rookout.Error2.deserializeBinaryFromReader(msg, reader);
};
/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.com.rookout.Error2} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.com.rookout.Error2}
 */


proto.com.rookout.Error2.deserializeBinaryFromReader = function (msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }

    var field = reader.getFieldNumber();

    switch (field) {
      case 1:
        var value =
        /** @type {string} */
        reader.readString();
        msg.setMessage(value);
        break;

      case 2:
        var value =
        /** @type {string} */
        reader.readString();
        msg.setType(value);
        break;

      case 3:
        var value = new proto.com.rookout.Variant2();
        reader.readMessage(value, proto.com.rookout.Variant2.deserializeBinaryFromReader);
        msg.setParameters(value);
        break;

      case 4:
        var value = new proto.com.rookout.Variant2();
        reader.readMessage(value, proto.com.rookout.Variant2.deserializeBinaryFromReader);
        msg.setExc(value);
        break;

      case 5:
        var value = new proto.com.rookout.Variant2();
        reader.readMessage(value, proto.com.rookout.Variant2.deserializeBinaryFromReader);
        msg.setTraceback(value);
        break;

      default:
        reader.skipField();
        break;
    }
  }

  return msg;
};
/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */


proto.com.rookout.Error2.prototype.serializeBinary = function () {
  var writer = new jspb.BinaryWriter();
  proto.com.rookout.Error2.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};
/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.com.rookout.Error2} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */


proto.com.rookout.Error2.serializeBinaryToWriter = function (message, writer) {
  var f = undefined;
  f = message.getMessage();

  if (f.length > 0) {
    writer.writeString(1, f);
  }

  f = message.getType();

  if (f.length > 0) {
    writer.writeString(2, f);
  }

  f = message.getParameters();

  if (f != null) {
    writer.writeMessage(3, f, proto.com.rookout.Variant2.serializeBinaryToWriter);
  }

  f = message.getExc();

  if (f != null) {
    writer.writeMessage(4, f, proto.com.rookout.Variant2.serializeBinaryToWriter);
  }

  f = message.getTraceback();

  if (f != null) {
    writer.writeMessage(5, f, proto.com.rookout.Variant2.serializeBinaryToWriter);
  }
};
/**
 * optional string message = 1;
 * @return {string}
 */


proto.com.rookout.Error2.prototype.getMessage = function () {
  return (
    /** @type {string} */
    jspb.Message.getFieldWithDefault(this, 1, "")
  );
};
/**
 * @param {string} value
 * @return {!proto.com.rookout.Error2} returns this
 */


proto.com.rookout.Error2.prototype.setMessage = function (value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};
/**
 * optional string type = 2;
 * @return {string}
 */


proto.com.rookout.Error2.prototype.getType = function () {
  return (
    /** @type {string} */
    jspb.Message.getFieldWithDefault(this, 2, "")
  );
};
/**
 * @param {string} value
 * @return {!proto.com.rookout.Error2} returns this
 */


proto.com.rookout.Error2.prototype.setType = function (value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};
/**
 * optional Variant2 parameters = 3;
 * @return {?proto.com.rookout.Variant2}
 */


proto.com.rookout.Error2.prototype.getParameters = function () {
  return (
    /** @type{?proto.com.rookout.Variant2} */
    jspb.Message.getWrapperField(this, proto.com.rookout.Variant2, 3)
  );
};
/**
 * @param {?proto.com.rookout.Variant2|undefined} value
 * @return {!proto.com.rookout.Error2} returns this
*/


proto.com.rookout.Error2.prototype.setParameters = function (value) {
  return jspb.Message.setWrapperField(this, 3, value);
};
/**
 * Clears the message field making it undefined.
 * @return {!proto.com.rookout.Error2} returns this
 */


proto.com.rookout.Error2.prototype.clearParameters = function () {
  return this.setParameters(undefined);
};
/**
 * Returns whether this field is set.
 * @return {boolean}
 */


proto.com.rookout.Error2.prototype.hasParameters = function () {
  return jspb.Message.getField(this, 3) != null;
};
/**
 * optional Variant2 exc = 4;
 * @return {?proto.com.rookout.Variant2}
 */


proto.com.rookout.Error2.prototype.getExc = function () {
  return (
    /** @type{?proto.com.rookout.Variant2} */
    jspb.Message.getWrapperField(this, proto.com.rookout.Variant2, 4)
  );
};
/**
 * @param {?proto.com.rookout.Variant2|undefined} value
 * @return {!proto.com.rookout.Error2} returns this
*/


proto.com.rookout.Error2.prototype.setExc = function (value) {
  return jspb.Message.setWrapperField(this, 4, value);
};
/**
 * Clears the message field making it undefined.
 * @return {!proto.com.rookout.Error2} returns this
 */


proto.com.rookout.Error2.prototype.clearExc = function () {
  return this.setExc(undefined);
};
/**
 * Returns whether this field is set.
 * @return {boolean}
 */


proto.com.rookout.Error2.prototype.hasExc = function () {
  return jspb.Message.getField(this, 4) != null;
};
/**
 * optional Variant2 traceback = 5;
 * @return {?proto.com.rookout.Variant2}
 */


proto.com.rookout.Error2.prototype.getTraceback = function () {
  return (
    /** @type{?proto.com.rookout.Variant2} */
    jspb.Message.getWrapperField(this, proto.com.rookout.Variant2, 5)
  );
};
/**
 * @param {?proto.com.rookout.Variant2|undefined} value
 * @return {!proto.com.rookout.Error2} returns this
*/


proto.com.rookout.Error2.prototype.setTraceback = function (value) {
  return jspb.Message.setWrapperField(this, 5, value);
};
/**
 * Clears the message field making it undefined.
 * @return {!proto.com.rookout.Error2} returns this
 */


proto.com.rookout.Error2.prototype.clearTraceback = function () {
  return this.setTraceback(undefined);
};
/**
 * Returns whether this field is set.
 * @return {boolean}
 */


proto.com.rookout.Error2.prototype.hasTraceback = function () {
  return jspb.Message.getField(this, 5) != null;
};
/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */


proto.com.rookout.Variant2.repeatedFields_ = [3, 4, 8, 9, 11];

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.com.rookout.Variant2.prototype.toObject = function (opt_includeInstance) {
    return proto.com.rookout.Variant2.toObject(opt_includeInstance, this);
  };
  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.com.rookout.Variant2} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */


  proto.com.rookout.Variant2.toObject = function (includeInstance, msg) {
    var f,
        obj = {
      variantTypeMaxDepth: jspb.Message.getFieldWithDefault(msg, 1, 0),
      originalTypeIndexInCache: jspb.Message.getFieldWithDefault(msg, 2, 0),
      attributeNamesInCacheList: (f = jspb.Message.getRepeatedField(msg, 3)) == null ? undefined : f,
      attributeValuesList: jspb.Message.toObjectList(msg.getAttributeValuesList(), proto.com.rookout.Variant2.toObject, includeInstance),
      originalSize: jspb.Message.getFieldWithDefault(msg, 5, 0),
      longValue: jspb.Message.getFieldWithDefault(msg, 6, 0),
      bytesIndexInCache: jspb.Message.getFieldWithDefault(msg, 7, 0),
      collectionKeysList: jspb.Message.toObjectList(msg.getCollectionKeysList(), proto.com.rookout.Variant2.toObject, includeInstance),
      collectionValuesList: jspb.Message.toObjectList(msg.getCollectionValuesList(), proto.com.rookout.Variant2.toObject, includeInstance),
      doubleValue: jspb.Message.getFloatingPointFieldWithDefault(msg, 10, 0.0),
      codeValuesList: jspb.Message.toObjectList(msg.getCodeValuesList(), variant_pb.Variant.CodeObject.toObject, includeInstance),
      timeValue: (f = msg.getTimeValue()) && google_protobuf_timestamp_pb.Timestamp.toObject(includeInstance, f),
      errorValue: (f = msg.getErrorValue()) && proto.com.rookout.Error2.toObject(includeInstance, f),
      complexValue: (f = msg.getComplexValue()) && variant_pb.Variant.Complex.toObject(includeInstance, f),
      livetail: (f = msg.getLivetail()) && variant_pb.Variant.LiveTailMessage.toObject(includeInstance, f)
    };

    if (includeInstance) {
      obj.$jspbMessageInstance = msg;
    }

    return obj;
  };
}
/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.com.rookout.Variant2}
 */


proto.com.rookout.Variant2.deserializeBinary = function (bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.com.rookout.Variant2();
  return proto.com.rookout.Variant2.deserializeBinaryFromReader(msg, reader);
};
/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.com.rookout.Variant2} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.com.rookout.Variant2}
 */


proto.com.rookout.Variant2.deserializeBinaryFromReader = function (msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }

    var field = reader.getFieldNumber();

    switch (field) {
      case 1:
        var value =
        /** @type {number} */
        reader.readUint32();
        msg.setVariantTypeMaxDepth(value);
        break;

      case 2:
        var value =
        /** @type {number} */
        reader.readUint32();
        msg.setOriginalTypeIndexInCache(value);
        break;

      case 3:
        var value =
        /** @type {!Array<number>} */
        reader.readPackedUint32();
        msg.setAttributeNamesInCacheList(value);
        break;

      case 4:
        var value = new proto.com.rookout.Variant2();
        reader.readMessage(value, proto.com.rookout.Variant2.deserializeBinaryFromReader);
        msg.addAttributeValues(value);
        break;

      case 5:
        var value =
        /** @type {number} */
        reader.readUint32();
        msg.setOriginalSize(value);
        break;

      case 6:
        var value =
        /** @type {number} */
        reader.readInt64();
        msg.setLongValue(value);
        break;

      case 7:
        var value =
        /** @type {number} */
        reader.readUint32();
        msg.setBytesIndexInCache(value);
        break;

      case 8:
        var value = new proto.com.rookout.Variant2();
        reader.readMessage(value, proto.com.rookout.Variant2.deserializeBinaryFromReader);
        msg.addCollectionKeys(value);
        break;

      case 9:
        var value = new proto.com.rookout.Variant2();
        reader.readMessage(value, proto.com.rookout.Variant2.deserializeBinaryFromReader);
        msg.addCollectionValues(value);
        break;

      case 10:
        var value =
        /** @type {number} */
        reader.readDouble();
        msg.setDoubleValue(value);
        break;

      case 11:
        var value = new variant_pb.Variant.CodeObject();
        reader.readMessage(value, variant_pb.Variant.CodeObject.deserializeBinaryFromReader);
        msg.addCodeValues(value);
        break;

      case 12:
        var value = new google_protobuf_timestamp_pb.Timestamp();
        reader.readMessage(value, google_protobuf_timestamp_pb.Timestamp.deserializeBinaryFromReader);
        msg.setTimeValue(value);
        break;

      case 16:
        var value = new proto.com.rookout.Error2();
        reader.readMessage(value, proto.com.rookout.Error2.deserializeBinaryFromReader);
        msg.setErrorValue(value);
        break;

      case 17:
        var value = new variant_pb.Variant.Complex();
        reader.readMessage(value, variant_pb.Variant.Complex.deserializeBinaryFromReader);
        msg.setComplexValue(value);
        break;

      case 18:
        var value = new variant_pb.Variant.LiveTailMessage();
        reader.readMessage(value, variant_pb.Variant.LiveTailMessage.deserializeBinaryFromReader);
        msg.setLivetail(value);
        break;

      default:
        reader.skipField();
        break;
    }
  }

  return msg;
};
/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */


proto.com.rookout.Variant2.prototype.serializeBinary = function () {
  var writer = new jspb.BinaryWriter();
  proto.com.rookout.Variant2.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};
/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.com.rookout.Variant2} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */


proto.com.rookout.Variant2.serializeBinaryToWriter = function (message, writer) {
  var f = undefined;
  f = message.getVariantTypeMaxDepth();

  if (f !== 0) {
    writer.writeUint32(1, f);
  }

  f = message.getOriginalTypeIndexInCache();

  if (f !== 0) {
    writer.writeUint32(2, f);
  }

  f = message.getAttributeNamesInCacheList();

  if (f.length > 0) {
    writer.writePackedUint32(3, f);
  }

  f = message.getAttributeValuesList();

  if (f.length > 0) {
    writer.writeRepeatedMessage(4, f, proto.com.rookout.Variant2.serializeBinaryToWriter);
  }

  f = message.getOriginalSize();

  if (f !== 0) {
    writer.writeUint32(5, f);
  }

  f = message.getLongValue();

  if (f !== 0) {
    writer.writeInt64(6, f);
  }

  f = message.getBytesIndexInCache();

  if (f !== 0) {
    writer.writeUint32(7, f);
  }

  f = message.getCollectionKeysList();

  if (f.length > 0) {
    writer.writeRepeatedMessage(8, f, proto.com.rookout.Variant2.serializeBinaryToWriter);
  }

  f = message.getCollectionValuesList();

  if (f.length > 0) {
    writer.writeRepeatedMessage(9, f, proto.com.rookout.Variant2.serializeBinaryToWriter);
  }

  f = message.getDoubleValue();

  if (f !== 0.0) {
    writer.writeDouble(10, f);
  }

  f = message.getCodeValuesList();

  if (f.length > 0) {
    writer.writeRepeatedMessage(11, f, variant_pb.Variant.CodeObject.serializeBinaryToWriter);
  }

  f = message.getTimeValue();

  if (f != null) {
    writer.writeMessage(12, f, google_protobuf_timestamp_pb.Timestamp.serializeBinaryToWriter);
  }

  f = message.getErrorValue();

  if (f != null) {
    writer.writeMessage(16, f, proto.com.rookout.Error2.serializeBinaryToWriter);
  }

  f = message.getComplexValue();

  if (f != null) {
    writer.writeMessage(17, f, variant_pb.Variant.Complex.serializeBinaryToWriter);
  }

  f = message.getLivetail();

  if (f != null) {
    writer.writeMessage(18, f, variant_pb.Variant.LiveTailMessage.serializeBinaryToWriter);
  }
};
/**
 * optional uint32 variant_type_max_depth = 1;
 * @return {number}
 */


proto.com.rookout.Variant2.prototype.getVariantTypeMaxDepth = function () {
  return (
    /** @type {number} */
    jspb.Message.getFieldWithDefault(this, 1, 0)
  );
};
/**
 * @param {number} value
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.setVariantTypeMaxDepth = function (value) {
  return jspb.Message.setProto3IntField(this, 1, value);
};
/**
 * optional uint32 original_type_index_in_cache = 2;
 * @return {number}
 */


proto.com.rookout.Variant2.prototype.getOriginalTypeIndexInCache = function () {
  return (
    /** @type {number} */
    jspb.Message.getFieldWithDefault(this, 2, 0)
  );
};
/**
 * @param {number} value
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.setOriginalTypeIndexInCache = function (value) {
  return jspb.Message.setProto3IntField(this, 2, value);
};
/**
 * repeated uint32 attribute_names_in_cache = 3;
 * @return {!Array<number>}
 */


proto.com.rookout.Variant2.prototype.getAttributeNamesInCacheList = function () {
  return (
    /** @type {!Array<number>} */
    jspb.Message.getRepeatedField(this, 3)
  );
};
/**
 * @param {!Array<number>} value
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.setAttributeNamesInCacheList = function (value) {
  return jspb.Message.setField(this, 3, value || []);
};
/**
 * @param {number} value
 * @param {number=} opt_index
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.addAttributeNamesInCache = function (value, opt_index) {
  return jspb.Message.addToRepeatedField(this, 3, value, opt_index);
};
/**
 * Clears the list making it empty but non-null.
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.clearAttributeNamesInCacheList = function () {
  return this.setAttributeNamesInCacheList([]);
};
/**
 * repeated Variant2 attribute_values = 4;
 * @return {!Array<!proto.com.rookout.Variant2>}
 */


proto.com.rookout.Variant2.prototype.getAttributeValuesList = function () {
  return (
    /** @type{!Array<!proto.com.rookout.Variant2>} */
    jspb.Message.getRepeatedWrapperField(this, proto.com.rookout.Variant2, 4)
  );
};
/**
 * @param {!Array<!proto.com.rookout.Variant2>} value
 * @return {!proto.com.rookout.Variant2} returns this
*/


proto.com.rookout.Variant2.prototype.setAttributeValuesList = function (value) {
  return jspb.Message.setRepeatedWrapperField(this, 4, value);
};
/**
 * @param {!proto.com.rookout.Variant2=} opt_value
 * @param {number=} opt_index
 * @return {!proto.com.rookout.Variant2}
 */


proto.com.rookout.Variant2.prototype.addAttributeValues = function (opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 4, opt_value, proto.com.rookout.Variant2, opt_index);
};
/**
 * Clears the list making it empty but non-null.
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.clearAttributeValuesList = function () {
  return this.setAttributeValuesList([]);
};
/**
 * optional uint32 original_size = 5;
 * @return {number}
 */


proto.com.rookout.Variant2.prototype.getOriginalSize = function () {
  return (
    /** @type {number} */
    jspb.Message.getFieldWithDefault(this, 5, 0)
  );
};
/**
 * @param {number} value
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.setOriginalSize = function (value) {
  return jspb.Message.setProto3IntField(this, 5, value);
};
/**
 * optional int64 long_value = 6;
 * @return {number}
 */


proto.com.rookout.Variant2.prototype.getLongValue = function () {
  return (
    /** @type {number} */
    jspb.Message.getFieldWithDefault(this, 6, 0)
  );
};
/**
 * @param {number} value
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.setLongValue = function (value) {
  return jspb.Message.setProto3IntField(this, 6, value);
};
/**
 * optional uint32 bytes_index_in_cache = 7;
 * @return {number}
 */


proto.com.rookout.Variant2.prototype.getBytesIndexInCache = function () {
  return (
    /** @type {number} */
    jspb.Message.getFieldWithDefault(this, 7, 0)
  );
};
/**
 * @param {number} value
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.setBytesIndexInCache = function (value) {
  return jspb.Message.setProto3IntField(this, 7, value);
};
/**
 * repeated Variant2 collection_keys = 8;
 * @return {!Array<!proto.com.rookout.Variant2>}
 */


proto.com.rookout.Variant2.prototype.getCollectionKeysList = function () {
  return (
    /** @type{!Array<!proto.com.rookout.Variant2>} */
    jspb.Message.getRepeatedWrapperField(this, proto.com.rookout.Variant2, 8)
  );
};
/**
 * @param {!Array<!proto.com.rookout.Variant2>} value
 * @return {!proto.com.rookout.Variant2} returns this
*/


proto.com.rookout.Variant2.prototype.setCollectionKeysList = function (value) {
  return jspb.Message.setRepeatedWrapperField(this, 8, value);
};
/**
 * @param {!proto.com.rookout.Variant2=} opt_value
 * @param {number=} opt_index
 * @return {!proto.com.rookout.Variant2}
 */


proto.com.rookout.Variant2.prototype.addCollectionKeys = function (opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 8, opt_value, proto.com.rookout.Variant2, opt_index);
};
/**
 * Clears the list making it empty but non-null.
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.clearCollectionKeysList = function () {
  return this.setCollectionKeysList([]);
};
/**
 * repeated Variant2 collection_values = 9;
 * @return {!Array<!proto.com.rookout.Variant2>}
 */


proto.com.rookout.Variant2.prototype.getCollectionValuesList = function () {
  return (
    /** @type{!Array<!proto.com.rookout.Variant2>} */
    jspb.Message.getRepeatedWrapperField(this, proto.com.rookout.Variant2, 9)
  );
};
/**
 * @param {!Array<!proto.com.rookout.Variant2>} value
 * @return {!proto.com.rookout.Variant2} returns this
*/


proto.com.rookout.Variant2.prototype.setCollectionValuesList = function (value) {
  return jspb.Message.setRepeatedWrapperField(this, 9, value);
};
/**
 * @param {!proto.com.rookout.Variant2=} opt_value
 * @param {number=} opt_index
 * @return {!proto.com.rookout.Variant2}
 */


proto.com.rookout.Variant2.prototype.addCollectionValues = function (opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 9, opt_value, proto.com.rookout.Variant2, opt_index);
};
/**
 * Clears the list making it empty but non-null.
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.clearCollectionValuesList = function () {
  return this.setCollectionValuesList([]);
};
/**
 * optional double double_value = 10;
 * @return {number}
 */


proto.com.rookout.Variant2.prototype.getDoubleValue = function () {
  return (
    /** @type {number} */
    jspb.Message.getFloatingPointFieldWithDefault(this, 10, 0.0)
  );
};
/**
 * @param {number} value
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.setDoubleValue = function (value) {
  return jspb.Message.setProto3FloatField(this, 10, value);
};
/**
 * repeated Variant.CodeObject code_values = 11;
 * @return {!Array<!proto.com.rookout.Variant.CodeObject>}
 */


proto.com.rookout.Variant2.prototype.getCodeValuesList = function () {
  return (
    /** @type{!Array<!proto.com.rookout.Variant.CodeObject>} */
    jspb.Message.getRepeatedWrapperField(this, variant_pb.Variant.CodeObject, 11)
  );
};
/**
 * @param {!Array<!proto.com.rookout.Variant.CodeObject>} value
 * @return {!proto.com.rookout.Variant2} returns this
*/


proto.com.rookout.Variant2.prototype.setCodeValuesList = function (value) {
  return jspb.Message.setRepeatedWrapperField(this, 11, value);
};
/**
 * @param {!proto.com.rookout.Variant.CodeObject=} opt_value
 * @param {number=} opt_index
 * @return {!proto.com.rookout.Variant.CodeObject}
 */


proto.com.rookout.Variant2.prototype.addCodeValues = function (opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 11, opt_value, proto.com.rookout.Variant.CodeObject, opt_index);
};
/**
 * Clears the list making it empty but non-null.
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.clearCodeValuesList = function () {
  return this.setCodeValuesList([]);
};
/**
 * optional google.protobuf.Timestamp time_value = 12;
 * @return {?proto.google.protobuf.Timestamp}
 */


proto.com.rookout.Variant2.prototype.getTimeValue = function () {
  return (
    /** @type{?proto.google.protobuf.Timestamp} */
    jspb.Message.getWrapperField(this, google_protobuf_timestamp_pb.Timestamp, 12)
  );
};
/**
 * @param {?proto.google.protobuf.Timestamp|undefined} value
 * @return {!proto.com.rookout.Variant2} returns this
*/


proto.com.rookout.Variant2.prototype.setTimeValue = function (value) {
  return jspb.Message.setWrapperField(this, 12, value);
};
/**
 * Clears the message field making it undefined.
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.clearTimeValue = function () {
  return this.setTimeValue(undefined);
};
/**
 * Returns whether this field is set.
 * @return {boolean}
 */


proto.com.rookout.Variant2.prototype.hasTimeValue = function () {
  return jspb.Message.getField(this, 12) != null;
};
/**
 * optional Error2 error_value = 16;
 * @return {?proto.com.rookout.Error2}
 */


proto.com.rookout.Variant2.prototype.getErrorValue = function () {
  return (
    /** @type{?proto.com.rookout.Error2} */
    jspb.Message.getWrapperField(this, proto.com.rookout.Error2, 16)
  );
};
/**
 * @param {?proto.com.rookout.Error2|undefined} value
 * @return {!proto.com.rookout.Variant2} returns this
*/


proto.com.rookout.Variant2.prototype.setErrorValue = function (value) {
  return jspb.Message.setWrapperField(this, 16, value);
};
/**
 * Clears the message field making it undefined.
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.clearErrorValue = function () {
  return this.setErrorValue(undefined);
};
/**
 * Returns whether this field is set.
 * @return {boolean}
 */


proto.com.rookout.Variant2.prototype.hasErrorValue = function () {
  return jspb.Message.getField(this, 16) != null;
};
/**
 * optional Variant.Complex complex_value = 17;
 * @return {?proto.com.rookout.Variant.Complex}
 */


proto.com.rookout.Variant2.prototype.getComplexValue = function () {
  return (
    /** @type{?proto.com.rookout.Variant.Complex} */
    jspb.Message.getWrapperField(this, variant_pb.Variant.Complex, 17)
  );
};
/**
 * @param {?proto.com.rookout.Variant.Complex|undefined} value
 * @return {!proto.com.rookout.Variant2} returns this
*/


proto.com.rookout.Variant2.prototype.setComplexValue = function (value) {
  return jspb.Message.setWrapperField(this, 17, value);
};
/**
 * Clears the message field making it undefined.
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.clearComplexValue = function () {
  return this.setComplexValue(undefined);
};
/**
 * Returns whether this field is set.
 * @return {boolean}
 */


proto.com.rookout.Variant2.prototype.hasComplexValue = function () {
  return jspb.Message.getField(this, 17) != null;
};
/**
 * optional Variant.LiveTailMessage livetail = 18;
 * @return {?proto.com.rookout.Variant.LiveTailMessage}
 */


proto.com.rookout.Variant2.prototype.getLivetail = function () {
  return (
    /** @type{?proto.com.rookout.Variant.LiveTailMessage} */
    jspb.Message.getWrapperField(this, variant_pb.Variant.LiveTailMessage, 18)
  );
};
/**
 * @param {?proto.com.rookout.Variant.LiveTailMessage|undefined} value
 * @return {!proto.com.rookout.Variant2} returns this
*/


proto.com.rookout.Variant2.prototype.setLivetail = function (value) {
  return jspb.Message.setWrapperField(this, 18, value);
};
/**
 * Clears the message field making it undefined.
 * @return {!proto.com.rookout.Variant2} returns this
 */


proto.com.rookout.Variant2.prototype.clearLivetail = function () {
  return this.setLivetail(undefined);
};
/**
 * Returns whether this field is set.
 * @return {boolean}
 */


proto.com.rookout.Variant2.prototype.hasLivetail = function () {
  return jspb.Message.getField(this, 18) != null;
};

goog.object.extend(exports, proto.com.rookout);
