// rookout/services/MemoryUsageLogService.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MemoryUsageLogService = void 0;

var _logger = require("../logger");

const config = require('../config');

const fs = require('fs');

const util = require('util');

const v8 = require('v8');

let log_fd = null;

class MemoryUsageLogService {
  constructor() {
    this.memoryUsageInterval = null;
  }

  start() {
    if (this.memoryUsageInterval) {
      return;
    }

    let log_file_path = config.MemoryUsageLogService.LOG_FILE;

    if (log_file_path !== null) {
      try {
        log_fd = fs.openSync(log_file_path, 'a');
      } catch (e) {
        console.log(util.format('Rookout - Error while trying to open a memory usage log file, file: %s, err: %s', log_file_path, e));
        return;
      }
    }

    if (config.MemoryUsageLogService.TRACE_GC_ENABLED) {
      this.enableGCTrace();
    }

    this.logMemoryUsageInterval = setInterval(this.logMemoryUsage, config.MemoryUsageLogService.INTERVAL_SECONDS * 1000).unref();
  }

  close() {
    if (config.MemoryUsageLogService.TRACE_GC_ENABLED) {
      this.disableGCTrace();
    }

    clearInterval(this.logMemoryUsageInterval);
    this.logMemoryUsageInterval = null;

    if (log_fd !== null) {
      fs.closeSync(log_fd);
      log_fd = null;
    }
  }

  logMemoryUsage() {
    let virtualMemory = process.memoryUsage();
    let heapStatistics = v8.getHeapStatistics();

    if (log_fd !== null) {
      let writeSyncOptions = {
        encoding: "utf8",
        flag: "a",
        mode: 0o666
      };
      fs.writeFileSync(log_fd, "Memory usage: " + JSON.stringify(virtualMemory) + "\n", writeSyncOptions);
      fs.writeFileSync(log_fd, "Heap statistics: " + JSON.stringify(heapStatistics) + "\n", writeSyncOptions);
    } else {
      _logger.logger.info("Memory usage: " + JSON.stringify(virtualMemory));

      _logger.logger.info("Heap statistics: " + JSON.stringify(heapStatistics));
    }
  }

  enableGCTrace() {
    v8.setFlagsFromString('--trace_gc');
  }

  disableGCTrace() {
    v8.setFlagsFromString('--notrace_gc');
  }

}

exports.MemoryUsageLogService = MemoryUsageLogService;
