// rookout/processor/namespaces/InspectorStackNamespace.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Namespace = _interopRequireDefault(require("./Namespace"));

var _ListNamespace = _interopRequireDefault(require("./ListNamespace"));

var _InspectorFrameNamespace = _interopRequireDefault(require("./InspectorFrameNamespace"));

var _InspectorTracebackNamespace = require("./InspectorTracebackNamespace");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DEFAULT_TRACEBACK_LIMIT = 1000;
const DEFAULT_FRAMES_LIMIT = 100; // Note - this is a serious duplication of the V8StackNamespace

class InspectorStackNamespace extends _Namespace.default {
  constructor(inspector, callFrames) {
    super();
    this.inspector = inspector;
    this.callFrames = callFrames;
  }

  readKey(key) {
    return new _InspectorFrameNamespace.default(this.inspector, this.callFrames[key]);
  }

  callMethod(name, args) {
    switch (name) {
      case "traceback":
        return this.traceback(args);

      case "frames":
        return this.frames(args);

      default:
        return super.callMethod(name, args);
    }
  }

  traceback(args) {
    // Decide on dump depth
    let depth = DEFAULT_TRACEBACK_LIMIT;

    if (null != args && '' !== args) {
      depth = parseInt(args);
    }

    if (depth > this.callFrames.length) {
      depth = this.callFrames.length;
    }

    return new _InspectorTracebackNamespace.InspectorTracebackNamespace(this.inspector, this.callFrames, depth);
  }

  frames(args) {
    // Decide on dump depth
    let depth = DEFAULT_FRAMES_LIMIT;

    if (null != args && '' !== args) {
      depth = parseInt(args);
    }

    if (depth > this.callFrames.length) {
      depth = this.callFrames.length;
    }

    let result = [];

    for (let i = 0; i < depth; ++i) {
      let frame = new _InspectorFrameNamespace.default(this.inspector, this.callFrames[i]);
      result.push(frame.callMethod('dump'));
    }

    return new _ListNamespace.default(result);
  }

}

exports.default = InspectorStackNamespace;
