// rookout/processor/namespaces/LogRecordNamespace.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _ContainerNamespace = _interopRequireDefault(require("./ContainerNamespace"));

var _JSObjectNamespace = _interopRequireDefault(require("./JSObjectNamespace"));

var _LogRecord = require("../../services/LogRecord");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class LogRecordNamespace extends _JSObjectNamespace.default {
  constructor(logRecord) {
    super(logRecord);
  }

  callMethod(name, args) {
    switch (name) {
      case "dump":
        return this.dump();

      case "format":
        return this.format();

      default:
        return new _ContainerNamespace.default();
    }
  }

  format() {
    return new _JSObjectNamespace.default(this.obj.getMessage());
  }

  readAttribute(name) {
    this.prepareLogRecordForReadingAttribute(name);
    return _JSObjectNamespace.default.prototype.readAttribute.call(this, name);
  }

  prepareLogRecordForReadingAttribute(name) {
    switch (name) {
      case _LogRecord.LOG_RECORD_MESSAGE_FIELD_NAME:
        this.obj.formatMessage();
        break;

      case _LogRecord.LOG_RECORD_LEVEL_NAME_FIELD_NAME:
        this.obj.formatLevelName();
        break;
    }
  }

  dump() {
    return new _ContainerNamespace.default({
      'msg': new _JSObjectNamespace.default(this.obj.getMessage()),
      'formatted_message': this.format(),
      'level_name': new _JSObjectNamespace.default(this.obj.getLevelName()),
      'time': new _JSObjectNamespace.default(this.obj.time),
      'module': new _JSObjectNamespace.default(this.obj.module) // todo: send more fields. (fields implemented in the python rook:)
      // 'name': new JSObjectNamespace(this.obj.name),
      // 'args': new JSObjectNamespace(this.obj.args),
      // 'level_no': new JSObjectNamespace(this.obj.levelno),
      // 'path_name': new JSObjectNamespace(this.obj.pathname),
      // 'filename': new JSObjectNamespace(this.obj.filename),
      // 'lineno': new JSObjectNamespace(this.obj.lineno),
      // 'function': new JSObjectNamespace(this.obj.funcName),
      // 'thread_id': new JSObjectNamespace(this.obj.thread),
      // 'thread_name': new JSObjectNamespace(this.obj.threadName),
      // 'process_name': new JSObjectNamespace(this.obj.processName),
      // 'process_id': new JSObjectNamespace(this.obj.process)

    });
  }

}

exports.default = LogRecordNamespace;
