// rookout/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MSToNS = MSToNS;
exports.NanosecToMillisec = NanosecToMillisec;
exports.StringUtils = void 0;
exports.canonizeFileName = canonizeFileName;
exports.conditionalRequire = conditionalRequire;
exports.emptyCallack = emptyCallack;
exports.getEnv = getEnv;
exports.getProcessTimeInMillisWithDecimal = getProcessTimeInMillisWithDecimal;
exports.getTimeInNanoseconds = getTimeInNanoseconds;

const path = require('path');

const process = require('process');

const NUM_OF_NANO_IN_SECONDS = 10 ** 9;

class StringUtils {
  static trim(string) {
    return string.replace(/^\s+|\s+$/g, '');
  }

  /***
   * Strip a string on both ends from specific chars
   * @param string
   * @param chars
   */
  static strip(string, chars) {
    const charsArray = [];

    for (let char of chars) {
      switch (char) {
        case "}":
        case "{":
        case "]":
        case "[":
          charsArray.push("\\" + char);
          break;

        default:
          charsArray.push(char);
      }
    }

    const re = RegExp(`^[${charsArray.join("")}]+|[${charsArray.join("")}]+$`, "g");
    return string.replace(re, '');
  }

}

exports.StringUtils = StringUtils;

function getEnv(envName, defaultValue) {
  return process.env[envName] || defaultValue;
}

function canonizeFileName(filename) {
  return path.normalize(filename.replace(/[\\\/]/g, '/'));
}

function getTimeInNanoseconds() {
  let hrtime_result = process.hrtime();
  let seconds = hrtime_result[0];
  let nanoseconds_remainder = hrtime_result[1];
  return seconds * NUM_OF_NANO_IN_SECONDS + nanoseconds_remainder;
}

function MSToNS(milliseconds) {
  return milliseconds * 1000000;
}

function NanosecToMillisec(nanoseconds) {
  return nanoseconds / 1000000;
}

function emptyCallack() {}

function getProcessTimeInMillisWithDecimal() {
  // process.hrtime() gives a tuple (seconds, nanoseconds) -> we convert to milliseconds with decimal
  const hrTime = process.hrtime();
  return hrTime[0] * 1000 + hrTime[1] / 1e6;
}

function conditionalRequire(filename) {
  try {
    return require(`${filename}`);
  } catch (err) {
    return undefined;
  }
}
